README.md
# HTTP MANAGER #

This is a little manager that allows web applications to gain access to an underlying OS in an easy way. With it, a web app
can launch processes, read their output, and list files in the file system.

## USAGE ##

The idea behing http_manager is that web apps can use *xmlhttprequest()* to do several actions. Specifically they can launch a background process, read its stdout and/or stderr outputs (both at its end, or while they are running), or list the files in a folder.

## AVAILABLE COMMANDS ##

* **POST /run_program** launches a program, with its name (and optional path), parameters and so on in the data block. This
  program is launched inside a *shell*, and will not block the http manager because it is launched as a child, so it is possible
  to add *shell things* like redirections, pipes and so on.
  
  This command will return a JSON structure with a single element, named *pid*, with the pid number of the process launched.
  
  **IMPORTANT** It is mandatory for this and the next commands to work, to have a working */bin/sh* binary in the system.

* **POST /run_program_with_pipes** works like the previous command, but will redirect both *stdout* and *stderr*, to make them
  available by using the *get_result* and *get_partial_result* commands.

* **GET /get_result?pid=PID** allows to get the status and (optionally) the output of a launched program. The *PID* parameter
  must contain the value returned by *run_program* or *run_program_with_pipes*. It returns a JSON structure with four elements:
  
  * *running* if it is *true*, the program is still running; if it is *false*, the program ended
  * *retval* only valid when *running* is *false*. It contains the return value given by the program
  * *stdout* if the program ended, it will contain all the output made to the standard output by the program; if not, it will be empty
  * *stderr* if the program ended, it will contain all the output made to the error output by the program; if not, it will be empty

* **GET /get_partial_result?pid=PID** it works like *get_result*, but *stdout* and *stderr* will contain all the data generated
  since the last *get_partial_result* executed (or since the begining, if none was executed). This allows to show the output of the
  program in realtime, instead of having to wait for it to end.

* **POST /get_files** returns the files available in the path sent in the data block. It returns them as a JSON array, with one
  one element per file. Also, each element is a dictionary with three elements: *filename*, with the name of the file, *date*, with
  the last modification date of the file, and *isdir*, which is a boolean that specifies if this file is a folder or a regular file.

## CONTACTING THE AUTHOR ##

Sergio Costas Rodriguez  
http://www.rastersoft.com  
raster@rastersoft.com  
rastersoft@gmail.com  



HISTORY.md
* Version 1.2 (2015/10/05)
  * All JSON parameters are in a single line to allow to replace "\n" with
    "\\\n" in the Javascript code, allowing that way multiline support
* Version 1.1 (2015/10/05)
  * Now only allows connections from localhost
* Version 1.0 (2015/10/05)
  * First public version


src/http.c
/*
 * Copyright 2011-15 Raster Software Vigo
 *
 * This file is part of http_manager, derived from series_rss
 *
 * series_rss is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * series_rss is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

/*
 * http.c
 *
 *  Created on: 19/12/2011
 *      Author: raster (http://www.rastersoft.com)
 */

#include <sys/socket.h>
#include <sys/types.h>
#include <netinet/in.h>
#include <netdb.h>
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <unistd.h>
#include <errno.h>
#include <signal.h>
#include <stdbool.h>

#include "http.h"

bool http_command_path_are(struct http_petition *object,char *command, char *path) {

	if (0 != strcmp(command,object->header_command)) {
		return false;
	}
	if (0 != strcmp(path,object->header_path)) {
		return false;
	}
	return true;
}

int http_receive_petition(int sockfd, struct http_petition *object) {

	int rss_block,tmp_size;
	int content_length=-1;
	int header_size=-1;
	char *head_line;

	if(object->header) {
		free(object->header);
	}
	object->header=NULL;
	object->total_size=0;

	rss_block=0;
	object->error=0;
	while((content_length==-1)||(header_size==-1)||((object->total_size-header_size)<content_length)) {
		if (object->total_size==rss_block) {
			rss_block+=RSS_BUFF_SIZE;
			head_line=realloc(object->header,rss_block+1);
			object->header=head_line;
		}
		tmp_size=read(sockfd,object->header+object->total_size,rss_block-object->total_size);
		if (tmp_size<=0) {
			object->header[object->total_size]=0;
			return -1;
		}
		object->total_size+=tmp_size;
		object->header[object->total_size]=0;
		if ((object->check_for_size)&&(content_length==-1)) {
			http_get_data(object,0);
			if (object->data_size!=-1) {
				head_line=strstr(object->header,"\r\nContent-Length: ");
				if (head_line) {
					sscanf(head_line+18,"%d",&content_length);
				} else {
					content_length=0;
				}
				header_size=object->total_size-object->data_size;
			}
		}
	}
	http_fill_header(object);
	return 0;
}

void http_free_header_params(struct http_petition *object) {

	if (object == NULL) {
		return;
	}

	struct http_header_param *c,*n;
	for(c=object->header_params;c != NULL; c = n) {
		n = c->next;
		free(c->key);
		free(c->value);
		free(c);
	}
	object->header_params = NULL;
}

char *add_header_param(struct http_petition *object,char *param_list) {

	char *pointer,*pointer2;
	struct http_header_param *param;

	pointer = strchr(param_list,'&');
	if (pointer != NULL) {
		*pointer = 0;
	}
	param = (struct http_header_param *)malloc(sizeof(struct http_header_param));
	if (param == NULL) {
		return NULL;
	}
	bzero(param,sizeof(struct http_header_param));

	pointer2 = strchr(param_list,'=');
	if (pointer2 == NULL) { // no value
		param->key = strdup(param_list);
	} else {
		*pointer2 = 0;
		param->key = strdup(param_list);
		param->value = strdup(pointer2 + 1);
		*pointer2 = '=';
	}
	param->next = object->header_params;
	object->header_params = param;
	if (pointer != NULL) {
		*pointer = '&';
	}
	return pointer;
}

void http_fill_header(struct http_petition *object) {

	char *pointer1, *pointer2, *pointer3;

	pointer1 = strchr(object->header,' ');
	if (pointer1 == NULL) {
		return;
	}

	pointer3 = strchr(pointer1 + 1,' ');
	if (pointer3 != NULL) {
		*pointer3 = 0;
	}
	free(object->header_command);
	*pointer1 = 0;
	object->header_command = strdup(object->header);
	*pointer1 = ' ';

	pointer2 = strchr(object->header,'?');
	if (pointer2 == NULL) {
		object->header_path = strdup(pointer1+1);
	} else {
		*pointer2 = 0;
		object->header_path = strdup(pointer1+1);
		*pointer2 = '?';
		do {
			pointer2 = add_header_param(object,pointer2 + 1);
		} while(pointer2 != NULL);
	}
	if (pointer3 != NULL) {
		*pointer3 = ' ';
	}
}

void http_get_data(struct http_petition *object, char modify) {

	/* Jump over the HTTP header and return a pointer to the data itself */

	char *p,*q;
	int size;

	if ((object==NULL)||(object->header==NULL)) {
		return;
	}

	object->data_size=-1;
	size=object->total_size;
	for(p=object->header;*p;p++) {
		size--;
		if (*p=='\r') {
			q=p+1;
			if ((*q==0)||(*q!='\n')) {
				continue;
			}
			q++;
			if ((*q==0)||(*q!='\r')) {
				continue;
			}
			if (modify) {
				*(p+2)=0; // end of string at the end of headers (but preserve the last carriage return)
			}
			object->data=p+4;
			object->data_size=size-3;

			return;
		}
		if (*p=='\n') {
			q=p+1;
			if ((*q==0)||(*q!='\n')) {
				continue;
			}
			if (modify) {
				*(q)=0; // end of string at the end of headers (but preserve the last carriage return)
			}
			object->data=p+2;
			object->data_size=size-1;

			return;
		}
	}
}

int http_get_error(struct http_petition *object) {

	int v;

	if (object==NULL) {
		return -1;
	}
	if (strncmp(object->header,"HTTP/1.",7)) {
		return -2;
	}

	sscanf(object->header+9,"%d",&v);
	return (v);
}

char *http_tobase64(char *input) {

	char *data="ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";
	char buffer[3];
	int size,counter,atend;
	char *p1,*p2,*p3;
	char *output;

	size=strlen(input);
	output=malloc(size*2);
	for(p1=input,p3=output;*p1;) {
		atend=0;
		for(p2=buffer,counter=0;counter<3;counter++,p2++) {
			*p2=*p1;
			if (*p1) {
				p1++;
			} else {
				atend++;
			}
		}
		*(p3++)=data[(int)((buffer[0]>>2)&0x3F)];
		*(p3++)=data[(int)((((buffer[0]<<4)&0x30)|((buffer[1]>>4)&0x0F)))];
		if (atend<2) {
			*(p3++)=data[(int)((((buffer[1]<<2)&0x3C)|((buffer[2]>>6)&0x03)))];
		} else {
			*(p3++)='=';
		}
		if (atend==0) {
			*(p3++)=data[(int)(buffer[2]&0x3F)];
		} else {
			*(p3++)='=';
		}
		*p3=0;
	}
	return output;
}

void http_free_petition(struct http_petition *object) {

	if (object==NULL) {
		return;
	}

	free(object->session_id);
	free(object->header);
	free(object->userpass);
	free(object->header_command);
	free(object->header_path);

	http_free_buffer(&object->header_buffer);
	http_free_buffer(&object->data_buffer);
	http_free_header_params(object);

	free(object);
}

void http_alarm_cb(int v) {

	printf("Timeout\n");
}


/********************************************************************
 * This function accepts a HTTP connection at the socket SOCKFD,    *
 * reads the header and, optionally, the data sent by the browser,  *
 * and returns it in the OBJECT structure. Also returns the new     *
 * socket.                                                          *
 ********************************************************************/

struct http_petition * http_accept_connection(int sockfd) {

	int newsockfd;
	socklen_t clen;
	struct sockaddr_in cli_addr;
	struct http_petition *object;
	char *p;

	object=malloc(sizeof(struct http_petition));
	if(object==NULL) {
		return NULL;
	}
	bzero(object,sizeof(struct http_petition));

	clen = sizeof(cli_addr);
	newsockfd = accept(sockfd, (struct sockaddr *) &cli_addr, &clen);
	if (newsockfd < 0) {
		printf("Can't accept connection\n");
		object->error=1;
		return object;
	}

	object->error=0;
	object->check_for_size=1;
	http_receive_petition(newsockfd,object);
	http_get_data(object,1);
	object->sockfd=newsockfd;
	object->http_version='0';
	p=strstr(object->header,"HTTP/1.");
	if (p!=NULL) {
		object->http_version=*(p+7);
	}
	return object;
}

void http_append_data(struct dinstr *object,char *string,int size) {

	if ((object==NULL)||(size==0)) {
		return;
	}

	while ((object->size+size+1)>object->fullsize) {
		object->fullsize+=RSS_BUFF_SIZE;
		object->buffer=realloc(object->buffer,object->fullsize);
	}

	memcpy(object->buffer+object->size,string,size);
	object->size+=size;
	*(object->buffer+object->size)=0;
}

void http_append_string(struct dinstr *object,char *string) {

	http_append_data(object,string,strlen(string));

}

void http_free_buffer(struct dinstr *object) {

	if (object == NULL) {
		return;
	}
	free(object->buffer);
	object->buffer=NULL;
	object->size=0;
	object->fullsize=0;
}

void http_send_end(struct http_petition *object) {

	char cadena[1024];
	if (object->data_buffer.size>0) {
		sprintf(cadena,"Content-Length: %d\r\n\r\n",object->data_buffer.size);
		http_append_string(&object->header_buffer,cadena);
	} else {
		http_append_string(&object->header_buffer,"\r\n");
	}

	//printf("Envio de vuelta: %s\n\n",object->header_buffer.buffer);

	write(object->sockfd,object->header_buffer.buffer,object->header_buffer.size);
	if (object->data_buffer.size>0) {
		write(object->sockfd,object->data_buffer.buffer,object->data_buffer.size);
		//printf("%s",object->data_buffer.buffer);
	}
	//printf("\n");
	http_free_buffer(&object->header_buffer);
	http_free_buffer(&object->data_buffer);
	close(object->sockfd);

}

void http_send_str(struct http_petition *object,char *str) {

	http_append_string(&object->data_buffer,str);

}

void http_send_data(struct http_petition *object, char *data, int size) {

	http_append_data(&object->data_buffer,data,size);

}


void http_send_header(struct http_petition *object, int error,char *str_error) {

	char str[256];

	if (object->http_version=='1') {
		sprintf(str,"HTTP/1.1 %d %s\r\nConnection: close\r\n",error,str_error);
	} else {
		sprintf(str,"HTTP/1.0 %d %s\r\n",error,str_error);
	}
	http_append_string(&object->header_buffer,str);
}

void http_send_header_var(struct http_petition *object, char *var) {

	http_append_string(&object->header_buffer,var);
	http_append_string(&object->header_buffer,"\r\n");

}


src/http.h
/*
 * Copyright 2011-15 Raster Software Vigo
 *
 * This file is part of http_manager, derived from series_rss
 *
 * series_rss is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * series_rss is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

/*
 * http.h
 *
 *  Created on: 19/12/2011
 *      Author: raster (http://www.rastersoft.com)
 */

#ifndef HTTP_H_
#define HTTP_H_

#define RSS_TIMEOUT 10
#define RSS_BUFF_SIZE 4096

struct dinstr {

	char *buffer;
	int size;
	int fullsize;
};

struct http_extra_data {
	char *user;
	char *pass;
	char *header;
	char *data;
	int data_size;
};

struct http_header_param {

	char *key;
	char *value;
	struct http_header_param *next;
};

struct http_petition {

	char *userpass;
	char *data;
	char *header;
	char *header_command;
	char *header_path;
	char *session_id;
	int error;
	int http_error;
	int data_size;
	int total_size;
	char check_for_size;
	int sockfd;
	char http_version;
	struct dinstr header_buffer;
	struct dinstr data_buffer;
	struct http_extra_data *extra;
	struct http_header_param *header_params;
};

void http_connect(char *url, struct http_petition *);
struct http_petition *http_accept_connection(int sockfd);
void http_send_header(struct http_petition *object, int error,char *str_error);
void http_send_header_var(struct http_petition *object, char *var);
void http_send_str(struct http_petition *object,char *str);
void http_send_data(struct http_petition *object, char *data, int size);
void http_send_end(struct http_petition *object);
void http_get_data(struct http_petition *, char modify);
int http_receive_petition(int sockfd, struct http_petition *object);

struct http_petition *http_do_petition(char *url, struct http_extra_data *extra);
void http_free_petition(struct http_petition *);

char *http_tobase64(char *input);

void http_free_buffer(struct dinstr *object);
void http_fill_header(struct http_petition *object);

bool http_command_path_are(struct http_petition *object,char *command, char *path);

#endif /* HTTP_H_ */


src/debug.c
/*
 * Copyright 2011-15 Raster Software Vigo
 *
 * This file is part of http_manager, derived from series_rss
 *
 * series_rss is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * series_rss is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

/*
 * debug.c
 *
 *  Created on: 5 de oct. de 2015
 *      Author: raster
 */

#include <stdio.h>

#include "debug.h"

static enum DEBUG_MODE level = DEBUG_WARNING;

void debug(enum DEBUG_MODE mode, char *msg) {

	if (mode < level) {
		return;
	}

	printf(msg);
}

void debug_str(enum DEBUG_MODE mode, char *msg,char *string) {
	if (mode < level) {
		return;
	}
	printf(msg,string);
}

void debug_int(enum DEBUG_MODE mode, char *msg,int value) {
	if (mode < level) {
		return;
	}
	printf(msg,value);
}

void set_debug_level(enum DEBUG_MODE mode) {
	level = mode;
}


src/debug.h
/*
 * Copyright 2011-15 Raster Software Vigo
 *
 * This file is part of http_manager, derived from series_rss
 *
 * series_rss is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * series_rss is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

/*
 * debug.h
 *
 *  Created on: 5 de oct. de 2015
 *      Author: raster
 */

#ifndef SRC_DEBUG_H_
#define SRC_DEBUG_H_

enum DEBUG_MODE {DEBUG_INFO, DEBUG_WARNING, DEBUG_ERROR, DEBUG_CRITICAL};

void debug(enum DEBUG_MODE, char *msg);
void debug_str(enum DEBUG_MODE, char *msg,char *string);
void debug_int(enum DEBUG_MODE, char *msg,int value);
void set_debug_level(enum DEBUG_MODE);

#endif /* SRC_DEBUG_H_ */


src/http_manager.c
/*
 * Copyright 2011-15 Raster Software Vigo
 *
 * This file is part of http_manager, derived from series_rss
 *
 * series_rss is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * series_rss is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

/*
 * http_manager.c
 *
 *      Author: raster (http://www.rastersoft.com)
 */

#define _BSD_SOURCE

//#define SCRIPTS_PATH "/home/raster/workspace/webtv2/bg_apps/etc/scripts"
#define SCRIPTS_PATH "/etc/scripts"

#include <stdbool.h>

#include <dirent.h>
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <sys/types.h>
#include <sys/wait.h>
#include <regex.h>
#include <signal.h>
#include <unistd.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <poll.h>
#include <sys/time.h>
#include <dirent.h>
#include <sys/stat.h>
#include <errno.h>
#include <netinet/in.h>
#include <arpa/inet.h>

#include "http_manager.h"
#include "http.h"
#include "debug.h"
#include "process_list.h"

int time_val;
int main_port;
volatile int do_exit;
volatile int do_reload;


void remove_slash_atend(char *cadena,int jump_start) {

	int size;

	size=strlen(cadena);
	if ((size>jump_start)&&(cadena[size-1]=='/')) {
		cadena[size-1]=0;
	}
}


void term_handler_cb(int v) {
	do_exit=1;
}

void run_external_program(struct http_petition *object,bool piping) {

	char *cadena;
	struct Pipe_element *pobject_out;
	struct Pipe_element *pobject_err;
	char cadena2[32];
	int pipefd_out[2];
	int pipefd_err[2];
	int pid;

	if (piping) {
		pipe(pipefd_out);
		pipe(pipefd_err);
	} else {
		pipefd_out[0] = 0;
		pipefd_err[0] = 0;
	}
	pobject_out = pipe_new(pipefd_out[0],PIPE_CHILD_STDOUT);
	pobject_err = pipe_new(pipefd_err[0],PIPE_CHILD_STDERR);
	if ((pobject_out == NULL) ||(pobject_err == NULL)) {
		pipe_delete(pobject_out);
		pipe_delete(pobject_err);
		if (piping) {
			close(pipefd_out[0]);
			close(pipefd_out[1]);
			close(pipefd_err[0]);
			close(pipefd_err[1]);
		}
		http_send_header(object,500,"Not enought memory");
		debug(DEBUG_ERROR,"Not enough memory for creating a new process in the system\n");
		return;
	}

	pid = fork();
	if (pid == -1) {
		pipe_delete(pobject_out);
		pipe_delete(pobject_err);
		if (piping) {
			close(pipefd_out[0]);
			close(pipefd_out[1]);
			close(pipefd_err[0]);
			close(pipefd_err[1]);
		}
		http_send_header(object,500,"Not enought memory");
		debug(DEBUG_ERROR,"Not enough memory for creating a child\n");
		return;
	}
	if (pid == 0) {
		int retval;
		if (piping) {
			dup2(pipefd_out[1],1);
			dup2(pipefd_err[1],2);
		}
		cadena=malloc(object->data_size+1);
		memcpy(cadena,object->data,object->data_size);
		*(cadena+object->data_size)=0;
		debug_str(DEBUG_INFO,"Launching %s\n",cadena);
		retval = system(cadena);
		free(cadena);
		close(pipefd_out[1]);
		close(pipefd_err[1]);
		exit(retval);
	}

	// parent
	pobject_out->pid = pid;
	pobject_err->pid = pid;
	http_send_header(object,200,"OK");
	http_send_header_var(object,"Access-Control-Allow-Origin: *");
	sprintf(cadena2,"{ \"pid\" : %d }",pid);
	http_send_str(object,cadena2);
	http_send_end(object);
	http_free_petition(object);
}


void get_program_result(struct http_petition *object, bool get_partial) {

	int pid;
	int retval, status;
	struct Pipe_element *element_out, *element_err, *e;

	struct http_header_param *param;

	pid = -1;
	for (param = object->header_params; param!= NULL; param = param->next) {
		if (0 == strcmp(param->key,"pid")) {
			if (param->value != NULL) {
				pid = atoi(param->value);
				break;
			}
		}
	}

	debug_int(DEBUG_INFO, "Pido PID %d\n",pid);
	element_out = NULL;
	element_err = NULL;
	for(e = pipe_list.next; e!= NULL; e=e->next) {
		if (e->pid == pid) {
			if (e->type == PIPE_CHILD_STDOUT) {
				element_out = e;
			} else if (e->type == PIPE_CHILD_STDERR) {
				element_err = e;
			}
			if ((element_out != NULL) && (element_err != NULL)) {
				break;
			}
		}
	}
	if (element_out != NULL) {
		http_send_header(object,200,"OK");
		http_send_header_var(object,"Content-Type: text/plain; charset=UTF-8");
		/* Esta entrada es necesaria para que funcione XMLHttpRequest */
		http_send_header_var(object,"Access-Control-Allow-Origin: *");
		status = waitpid(pid,&retval,WNOHANG);
		char cadena[2048];
		sprintf(cadena,"{ \"running\" : %s, \"retval\" : %d, \"stdout\" : \"",status <= 0 ? "true" : "false",retval);
		http_send_str(object,cadena);
		if ((status > 0) || (get_partial)) {
			if (element_out->data != NULL) {
				http_send_data(object,element_out->data,element_out->data_size);
			}
		}
		http_send_str(object,"\", \"stderr\" : \"");
		if ((status > 0) || (get_partial)) {
			if (element_err->data != NULL) {
				http_send_data(object,element_err->data,element_err->data_size);
			}
		}
		http_send_str(object,"\"}");
		if (status > 0) {
			element_out->to_delete = true;
			element_err->to_delete = true;
		}
		if (get_partial) {
			pipe_reset_data(element_out);
			pipe_reset_data(element_err);
		}
	} else {
		debug(DEBUG_ERROR, "Asking for not valid PID\n");
		http_send_header(object,503,"PID not valid");
	}

	http_send_end(object);
	http_free_petition(object);
}

void get_files(struct http_petition *object) {

	DIR *dlist;
	struct dirent *dent;
	char *path,first,*path2,tmpstr[50];
	struct stat datos;
	int path_size,file_size;

	http_send_header(object,200,"OK");
	http_send_header_var(object,"Content-Type: text/plain; charset=UTF-8");
	http_send_header_var(object,"Access-Control-Allow-Origin: *");

	path=malloc(object->data_size+2);
	if (path==NULL) {
		dlist=NULL;
	} else {
		memcpy(path,object->data,object->data_size);
		*(path+object->data_size)=0;
		path_size=object->data_size;
		dlist=opendir(path);
		if ((path_size==0) || (path[path_size-1]!='/')) {
			path[path_size]='/';
			path[path_size+1]=0;
			path_size++;
		}
	}
	if (dlist==NULL) {
		http_send_str(object,"{ status: -1, files : []}");
	} else {
		http_send_str(object,"{ status: 0 , files: [");
		first=0;
		do {
			dent=readdir(dlist);
			if (dent!=NULL) {
				if ((dent->d_type!=DT_REG)&&(dent->d_type!=DT_DIR)) {
					continue;
				}
				if (dent->d_name[0]=='.') {
					continue;
				}
				file_size=strlen(dent->d_name);
				path2=malloc(path_size+file_size+1);
				sprintf(path2,"%s%s",path,dent->d_name);
				lstat(path2,&datos);
				if (first) {
					http_send_str(object,", ");
				}
				first=1;
				http_send_str(object,"{ filename:\"");
				http_send_str(object,dent->d_name);
				sprintf(tmpstr,"\", date: %lld , isdir: %s}",(long long int)datos.st_mtime,(dent->d_type==DT_DIR) ? "true" : "false");
				http_send_str(object,tmpstr);
				free(path2);
			}
		} while(dent);
		http_send_str(object,"]}");
		closedir(dlist);
	}
	free(path);
	http_send_end(object);
	http_free_petition(object);

}

void get_services(struct http_petition *object) {

	char first;

	char cadena2[8192];
	char cadena3[8192];
	DIR *folder;
	struct dirent *entrada;
	struct stat estado;

	http_send_header(object,200,"OK");
	http_send_header_var(object,"Content-Type: text/plain; charset=UTF-8");
	/* Esta entrada es necesaria para que funcione XMLHttpRequest */
	http_send_header_var(object,"Access-Control-Allow-Origin: *");

	folder=opendir(SCRIPTS_PATH);

	if (folder==NULL) {
		http_send_str(object,"{ status: -1, services : []}");
	} else {
		http_send_str(object,"{ status: 0, services : [");
		first=0;
		while((entrada=readdir(folder))!=NULL) {
			if (entrada->d_name[0]=='.') {
				continue;
			}
			if (0==strcmp(entrada->d_name,"tolaunch")) {
				continue;
			}
			if (first) {
				http_send_str(object,", ");
			}
			first=1;
			sprintf(cadena2,"%s/tolaunch/%s",SCRIPTS_PATH,entrada->d_name);
			sprintf(cadena3,"{ service: \"%s\", status: %d }",entrada->d_name,(stat(cadena2,&estado)==0) ? 1 : 0);
			http_send_str(object,cadena3);
		}
		http_send_str(object,"]}");
		closedir(folder);
	}
	http_send_end(object);
	http_free_petition(object);
}


void accept_connection(int sockfd) {

	struct http_petition *object;

	debug(DEBUG_INFO, "New conection\n");
	object=http_accept_connection(sockfd);
	if (object==NULL) {
		debug(DEBUG_ERROR, "Can't open a new conection\n");
		return;
	}
	if (object->error==0) {
		debug_str(DEBUG_INFO,"Command: %s\n",object->header_command);
		debug_str(DEBUG_INFO,"Path: %s\n",object->header_path);
		if (http_command_path_are(object,"POST", "/run_program")) {
			run_external_program(object,false);
		} else if (http_command_path_are(object,"POST", "/run_program_with_pipes")) {
			run_external_program(object,true);
		} else if (http_command_path_are(object,"GET","/get_result")) {
			get_program_result(object,false);
		} else if (http_command_path_are(object,"GET","/get_partial_result")) {
			get_program_result(object,true);
		} else if (http_command_path_are(object,"POST","/get_files")) {
			get_files(object);
		} else if (http_command_path_are(object,"GET","/get_services")) {
			get_services(object);
		} else {
			http_send_header(object,501,"COMMAND NOT VALID");
			http_send_header_var(object,"Content-Type: text/plain; charset=UTF-8");
			/* Esta entrada es necesaria para que funcione XMLHttpRequest */
			http_send_header_var(object,"Access-Control-Allow-Origin: *");
			http_send_end(object);
			http_free_petition(object);
		}
	} else {
		debug_int(DEBUG_ERROR, "Error when accepting the conection\n",object->error);
		http_send_header(object,500,"INTERNAL ERROR");
		http_send_header_var(object,"Content-Type: text/plain; charset=UTF-8");
		/* Esta entrada es necesaria para que funcione XMLHttpRequest */
		http_send_header_var(object,"Access-Control-Allow-Origin: *");
		http_send_end(object);
		http_free_petition(object);
	}
}

void read_data(struct Pipe_element *object) {

	int size;
	char buffer[8192];

	size = read(object->fd,buffer,8192);
	debug_int(DEBUG_INFO, "Read %d bytes\n",size);
	printf("Leido %d datos\n",size);
	if (size <= 0) {
		debug(DEBUG_ERROR, "Error while reading data\n");
		return;
	}
	if (object->data == NULL) {
		object->data = (char *)malloc(8192);
		object->data_size = 0;
		object->data_block_size = 8192;
	} else if ((object->data_block_size - object->data_size) < size) {
		object->data_block_size += 8192;
		object->data = realloc(object->data,object->data_block_size);
	}
	memcpy(object->data+object->data_size,buffer,size);
	object->data_size += size;
}

void do_loop() {


	int c,l;
	struct Pipe_element *e,*f;
	struct sockaddr_in serv_addr;

	pipe_list.fd = socket(AF_INET, SOCK_STREAM, 0);
	if (pipe_list.fd < 0) {
		debug(DEBUG_CRITICAL, "Error while creating the main socket\n");
		exit(-1);
	} else {
		bzero((char *) &serv_addr, sizeof(serv_addr));
		serv_addr.sin_family = AF_INET;
		serv_addr.sin_port = htons(main_port);
		serv_addr.sin_addr.s_addr = inet_addr("127.0.0.1");
		if (bind(pipe_list.fd, (struct sockaddr *) &serv_addr, sizeof(serv_addr)) < 0) {
			perror("Can't bind socket\n");
			close(pipe_list.fd);
			pipe_list.fd=0;
			exit(-1);
		} else {
			if ( 0!= listen(pipe_list.fd,5)) {
				printf("Can't listen on socket\n");
				close(pipe_list.fd);
				pipe_list.fd=0;
				exit(-1);
			}
		}
	}

	while(!do_exit) {
		for (e = &pipe_list, c = 0; e != NULL; e = e->next, c++) {
			fds[c].fd=e->fd;
			fds[c].events=POLLIN;
			fds[c].revents=0;
		}
		if (0 != poll(fds,pipe_list_size,-1)) {
			l = pipe_list_size;
			for (c=0;c<l;c++) {
				if (fds[c].revents & POLLIN) {
					if (fds[c].fd == pipe_list.fd) {
						accept_connection(fds[c].fd);
					} else {
						for (e = pipe_list.next; e != NULL; e = e->next) {
							if (e->fd == fds[c].fd) {
								read_data(e);
							}
						}
					}
				}
			}
			for (e = pipe_list.next; e != NULL;e = f) {
				f = e->next;
				if(e->to_delete) {
					pipe_delete(e);
				}
			}
		}
	}
	close(pipe_list.fd);
}

int main(int argc, char **argv) {

	int loop,t;
	struct sigaction term_action;
	struct sigaction quit_action;
	struct sigaction int_action;

	main_port = 9089;
	pipe_list.fd = 0; // default port
	pipe_list.next = NULL;
	pipe_list.prev = NULL;
	pipe_list.type = PIPE_MASTER;
	pipe_list_last = &pipe_list;
	pipe_list_size = 1;
	fds = (struct pollfd *)malloc(pipe_list_size * sizeof(struct pollfd));

	for(loop=1;loop<argc;loop++) {

		if (!strncmp(argv[loop],"-h",2)) { // help
			printf("\n  HTTP_MANAGER 1.1\n\nUsage: http_manager [-h] [-Pport]\n\n");
			printf("-h: shows this help.\n");
			printf("-P port to get/set configuration (default 9089).\n");
			exit(0);
		}

		if (!strncmp(argv[loop],"-P",2)) { // port
			t=atoi(2+argv[loop]);
			if ((t>=1)&&(t<=65535)) {
				main_port=t;
			}
			continue;
		}
	}

	do_exit=0;

	term_action.sa_handler=term_handler_cb;
	term_action.sa_flags=0;
	sigemptyset(&term_action.sa_mask);
	sigaction(SIGTERM,&term_action,NULL);

	quit_action.sa_handler=term_handler_cb;
	quit_action.sa_flags=0;
	sigemptyset(&quit_action.sa_mask);
	sigaction(SIGQUIT,&quit_action,NULL);

	int_action.sa_handler=term_handler_cb;
	int_action.sa_flags=0;
	sigemptyset(&int_action.sa_mask);
	sigaction(SIGINT,&int_action,NULL);

	do_loop();

	return 0;
}
